File: group 2/variant15/travel_booking_service.js
"use strict";

class Request {
    constructor(travelers, daysBefore, premium, flexible) {
        this.travelers = travelers;
        this.daysBefore = daysBefore;
        this.premium = premium;
        this.flexible = flexible;
        Object.freeze(this);
    }
}

class Result {
    constructor(status, travelers = 0, discountPercent = 0, options = 0, lateFee = 0, total = 0) {
        this.status = status;
        this.travelers = travelers;
        this.discountPercent = discountPercent;
        this.options = options;
        this.lateFee = lateFee;
        this.total = total;
        Object.freeze(this);
    }
}

class RoutePolicy {
    isOpen() {
        throw new Error("Потрібно надати залежність: RoutePolicy");
    }
}

class TravelCapacity {
    freeSeats() {
        throw new Error("Потрібно надати залежність: TravelCapacity");
    }
}

class TravelRepository {
    save(request, result) {
        throw new Error("Потрібно надати залежність: TravelRepository");
    }
}

function checkRange(value, minimum, maximum, field) {
    if (!Number.isSafeInteger(value) || value < minimum || value > maximum) {
        throw new RangeError("Недопустиме значення: " + field);
    }
}

function checkFlags(...values) {
    if (values.some(value => typeof value !== "boolean")) {
        throw new TypeError("Ознаки мають бути логічними значеннями");
    }
}

function readFlag(value, field) {
    if (typeof value !== "boolean") {
        throw new Error("Некоректні відомості: " + field);
    }
    return value;
}

function readCount(value, field) {
    if (!Number.isSafeInteger(value) || value < 0) {
        throw new Error("Некоректні відомості: " + field);
    }
    return value;
}

class TravelBookingService {
    constructor(route, capacity, bookings) {
        if (route == null || capacity == null || bookings == null) {
            throw new TypeError("Усі залежності мають бути задані");
        }
        this.route = route;
        this.capacity = capacity;
        this.bookings = bookings;
    }

    process(request) {
        if (!(request instanceof Request)) {
            throw new TypeError("Заявку не задано або її формат некоректний");
        }
        checkRange(request.travelers, 1, 6, "Кількість туристів");
        checkRange(request.daysBefore, 1, 179, "Дні до виїзду");
        checkFlags(request.premium, request.flexible);
        if (!readFlag(this.route.isOpen(), "Відкритість напрямку")) {
            return new Result("Напрямок закритий");
        }
        const seats = readCount(this.capacity.freeSeats(), "Вільні місця");
        if (seats < request.travelers) {
            return new Result("Недостатньо місць");
        }
        let optionRate;
        if (request.premium && request.flexible) {
            optionRate = 45_000;
        } else if (request.premium) {
            optionRate = 25_000;
        } else if (request.flexible) {
            optionRate = 20_000;
        } else {
            optionRate = 0;
        }
        const base = request.travelers * 100_000;
        const discount = request.daysBefore > 60 ? 10 : 0;
        const options = request.travelers * optionRate;
        const lateFee = request.daysBefore < 7 ? request.travelers * 15_000 : 0;
        const total = base - Math.floor(base * discount / 100) + options + lateFee;
        const result = new Result("Заброньовано", request.travelers, discount, options, lateFee, total);
        this.bookings.save(request, result);
        return result;
    }
}

module.exports = { Request, Result, RoutePolicy, TravelCapacity, TravelRepository, TravelBookingService };

File: group 3/variant15/TravelBookingService.java
public final class TravelBookingService {
    public record Request(int travelers, int daysBefore, boolean premium, boolean flexible) {}

    public record Result(String status, int travelers, int discountPercent, long options, long lateFee, long total) {}

    public interface RoutePolicy {
        boolean isOpen();
    }

    public interface TravelCapacity {
        int freeSeats();
    }

    public interface TravelRepository {
        void save(Request request, Result result);
    }

    private final RoutePolicy route;
    private final TravelCapacity capacity;
    private final TravelRepository bookings;

    public TravelBookingService(RoutePolicy route, TravelCapacity capacity, TravelRepository bookings) {
        this.route = java.util.Objects.requireNonNull(route, "Залежність не задано: route");
        this.capacity = java.util.Objects.requireNonNull(capacity, "Залежність не задано: capacity");
        this.bookings = java.util.Objects.requireNonNull(bookings, "Залежність не задано: bookings");
    }

    public Result process(Request request) {
        if (request == null) {
            throw new IllegalArgumentException("Заявку не задано");
        }
        checkRange(request.travelers(), 1, 6, "Кількість туристів");
        checkRange(request.daysBefore(), 1, 179, "Дні до виїзду");
        if (!route.isOpen()) {
            return rejected("Напрямок закритий");
        }
        int seats = capacity.freeSeats();
        if (seats < 0) {
            throw new IllegalStateException("Від’ємна кількість вільних місць");
        }
        if (seats < request.travelers()) {
            return rejected("Недостатньо місць");
        }
        long optionRate;
        if (request.premium() && request.flexible()) {
            optionRate = 45_000;
        } else if (request.premium()) {
            optionRate = 25_000;
        } else if (request.flexible()) {
            optionRate = 20_000;
        } else {
            optionRate = 0;
        }
        long base = request.travelers() * 100_000L;
        int discount = request.daysBefore() > 60 ? 10 : 0;
        long options = request.travelers() * optionRate;
        long lateFee = request.daysBefore() < 7 ? request.travelers() * 15_000L : 0;
        long total = base - base * discount / 100 + options + lateFee;
        Result result = new Result("Заброньовано", request.travelers(), discount, options, lateFee, total);
        bookings.save(request, result);
        return result;
    }

    private static Result rejected(String status) {
        return new Result(status, 0, 0, 0, 0, 0);
    }

    private static void checkRange(int value, int minimum, int maximum, String field) {
        if (value < minimum || value > maximum) {
            throw new IllegalArgumentException("Недопустиме значення: " + field);
        }
    }
}
